lomiri: init basic development stuff

OPNA2608 opened the pull request:
###### Description of changes

Working towards #99090, in more digestible steps than last time. This one is abit bigger, but the followups can be done more or less package-by-package with all of these present.

- create a maintainers team & package scope for Lomiri
  - if anyone wishes to join, add yourself and/or let me know :slightly_smiling_face:
- init Lomiri-belonging things
  - [cmake-extras](https://gitlab.com/ubports/development/core/cmake-extras): A set of CMake modules commonly used by Lomiri packages
  - [lomiri-api](https://gitlab.com/ubports/development/core/lomiri-api): An API for integrating with the Lomiri shell
  - [gmenuharness](https://gitlab.com/ubports/development/core/gmenuharness): A library to test GLib GMenuModels
- init some non-Lomiri things that are commonly used to drive tests
  - [libqtdbustest](https://launchpad.net/libqtdbustest): Library to test D-Bus interactions with Qt
  - [libqtdbusmock](https://launchpad.net/libqtdbusmock): Library to mock D-Bus interactions with Qt
  - [dbus-test-runner](https://launchpad.net/dbus-test-runner): Tool to run tasks under independent D-Bus sessions

~~The last 2 commits I still need to test on my system with all the Lomiri stuff, just threw them in here so those packages are less stale on arrival.~~

Working fine, keeping them in.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

OPNA2608 committed:
maintainers: Add Lomiri team

OPNA2608 committed:
lomiri.cmake-extras: init at unstable-2022-11-21

OPNA2608 committed:
dbus-test-runner: init at unstable-2019-10-02

OPNA2608 committed:
libqtdbustest: init at unstable-2017-01-06

OPNA2608 committed:
libqtdbusmock: init at unstable-2017-03-16

OPNA2608 committed:
lomiri.lomiri-api: init at 0.2.0

OPNA2608 committed:
lomiri.gmenuharness: init at 0.1.4

OPNA2608 committed:
lomiri.lomiri-api: 0.2.0 -> 0.2.1

OPNA2608 committed:
lomiri.gmenuharness: Fetch widget attribute rename patch

SuperSandro2000 commented on pkgs/development/tools/dbus-test-runner/default.nix:
```diff
@@ -0,0 +1,80 @@
+    description = "A small little utility to run a couple of executables under a new DBus session for testing";
+    homepage = "https://launchpad.net/dbus-test-runner";
+    license = licenses.gpl3Only;
+    platforms = platforms.all;
```
Wouldn't unix be a better fit?

OPNA2608 replied:
Everything that uses D-Bus has been marked `unix` now.

SuperSandro2000 reviewed (commented):
not sure on the package set if it is done correct for cross

OPNA2608 commented:
I'm not sure if the package set is correctly written to support cross, but most of the packages in it will use our `qtbase` where cross is borked anyway: #227900

OPNA2608 committed:
lomiri.cmake-extras: Change QML module installation path

We sometimes place QML modules into the bin output of derivations. Using CMAKE_INSTALL_BINDIR and backtracking from it
will ensure that the modules will end up under the bin output prefix, if it is enabled.

OPNA2608 commented on pkgs/desktops/lomiri/development/cmake-extras/default.nix:
```diff
@@ -0,0 +1,49 @@
+
+    # This is in a function that reverse dependencies use to determine where to install their files to
+    substituteInPlace src/QmlPlugins/QmlPluginsConfig.cmake \
+      --replace "\''${CMAKE_INSTALL_LIBDIR}/qt5/qml" "\''${CMAKE_INSTALL_PREFIX}/${qtbase.qtQmlPrefix}"
```
Tested the chain of dependencies up to `morph-browser`, with no change to outputs and with `lomiri-ui-toolkit` split into `out` and `bin` outputs. Adding `bin` outputs to at least some reverse dependencies that provide QML modules seems doable in the long run, so I've added this change.

OPNA2608 commented:
@SuperSandro2000 I've tried locally adding something to the package set that doesn't need Qt & building it with `--argstr crossSystem aarch64-linux` and that works fine, so I think the package set is done correct for cross. Bump.

wineee commented on pkgs/desktops/lomiri/development/cmake-extras/default.nix:
```diff
@@ -0,0 +1,49 @@
+
+    # This is in a function that reverse dependencies use to determine where to install their files to
+    substituteInPlace src/QmlPlugins/QmlPluginsConfig.cmake \
+      --replace "\''${CMAKE_INSTALL_LIBDIR}/qt5/qml" "\''${CMAKE_INSTALL_BINDIR}/../${qtbase.qtQmlPrefix}"
```
```suggestion
      --replace "\''${CMAKE_INSTALL_LIBDIR}/qt5/qml" "\''${CMAKE_INSTALL_PREFIX}/${qtbase.qtQmlPrefix}"
```

wineee replied:
Using CMAKE_INSTALL_BINDIR should also work. Actually, I prefer to set `QMLFILES_DESTINATION` in `cmakeFlags`  when we use cmake-extras

OPNA2608 replied:
Using `CMAKE_INSTALL_PREFIX` as the default and overriding `QMLFILES_DESTINATION` in derivations that want them in another output seems like an okay compromise.

OPNA2608 committed:
lomiri.cmake-extras: Change default QML location back to CMAKE_INSTALL_PREFIX

Going `${CMAKE_INSTALL_BINDIR}/../` looks weird. Derivations that want them in a non-`out` output can just set the `QMLFILES_DESTINATION` CMake flag as needed.

Co-authored-by: rewine <luhongxu@deepin.org>

SuperSandro2000 commented:
> @SuperSandro2000 Is the uncertainty about the package set's cross-buildability a blocker for you (and thus blocked by `qtbase` itself having borked cross) or is there anything else you'd like me to change?

Na, I was just curious and wanted to let you know that I do not know.

SuperSandro2000 commented:
If we missed something especially cross related we can fix that in another PR to not delay this one any longer.
